packages/xod-client/src/editor/components/Suggester.jsx

```diff
@@ -0,0 +1,184 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import classNames from 'classnames';
+import Autosuggest from 'react-autosuggest';
+import CustomScroll from 'react-custom-scroll';
+import Highlighter from 'react-highlight-words';
+
+import { KEYCODE } from '../../utils/constants';
+
+const getSuggestionValue = ({ item }) => item.path;
+
+class Suggester extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      value: '',
+      suggestions: [],
+    };
+
+    this.input = null;
+
+    this.renderItem = this.renderItem.bind(this);
+    this.onChange = this.onChange.bind(this);
+    this.onSuggestionsFetchRequested = this.onSuggestionsFetchRequested.bind(this);
+    this.onSuggestionsClearRequested = this.onSuggestionsClearRequested.bind(this);
+    this.onSuggestionSelected = this.onSuggestionSelected.bind(this);
+    this.storeInputReference = this.storeInputReference.bind(this);
+  }
+
+  componentDidMount() {
+    if (this.input) {
+      // A hack to avoid typing "i" into input when pressing Hotkey
+      setTimeout(() => this.input.focus(), 1);
+    }
+  }
+
+  onChange(e, { newValue, method }) {
+    if (
+      method === 'up' ||
+      method === 'down' ||
+      method === 'click'
+    ) {
+      return;
+    }
+
+    this.setState({ value: newValue });
+  }
+
+  onSelect(value) {
+    this.props.onAddNode(value);
+  }
+
+  onSuggestionsFetchRequested({ value, reason }) {
+    if (reason === 'suggestion-selected') return;
+
+    this.setState({
+      suggestions: this.getSuggestions(value),
+    });
+  }
+
+  onSuggestionsClearRequested() {
+    this.setState({
+      suggestions: [],
+    });
+  }
+
+  onSuggestionSelected(event, { suggestionValue, method }) {
+    if (method === 'enter') {
+      this.onSelect(suggestionValue);
+    }
+  }
+
+  getSuggestions(value) {
+    const { index } = this.props;
+    const inputValue = value.trim().toLowerCase();
+
+    if (inputValue.length === 0) { return []; }
+
+    const result = index.search(inputValue);
+    const filtered = result.filter(res => res.score < 0.2);
```

Comment:
Isn’t `0.2` is a view-level override of the index `threshold` option?

In any case, I’d abstract out the `search` function and place it in `xod-patch-search`. So that if we’d like to hack or even change the search engine, its consumers will not be affected.